File: go/2021/aug/count-good-nodes-in-binary-tree.go
package aug

import "leetcode/common"

func goodNodes(root *common.TreeNode) int {
    return goodNodes_(root, -10000)
}

func goodNodes_(root *common.TreeNode, ma int) int {
    if root == nil {
        return 0
    }
    var res int
    if root.Val >= ma {
        res = 1
    } else {
        res = 0
    }
    res += goodNodes_(root.Left, max(ma, root.Val))
    res += goodNodes_(root.Right, max(ma, root.Val))

    return res
}

File: java/src/main/java/cn/az/code/year2021/aug/CountGoodNodesInBinaryTree.java
package cn.az.code.year2021.aug;

import cn.az.code.common.TreeNode;

/**
 * @author az
 * @since 2021-08-17
 */
public class CountGoodNodesInBinaryTree {

    public int goodNodes(TreeNode root) {
        return goodNodes(root, -10000);
    }

    public int goodNodes(TreeNode root, int ma) {
        if (root == null) return 0;
        int res = root.val >= ma ? 1 : 0;
        res += goodNodes(root.left, Math.max(ma, root.val));
        res += goodNodes(root.right, Math.max(ma, root.val));
        return res;
    }
}
